src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,150 @@
+            if (arrayCreationExpression.Syntax.Ancestors().OfType<AttributeSyntax>().Any())
+                return;
+
+            // Can't use a UTF8 string inside an expression tree.
+            var semanticModel = context.Operation.SemanticModel;
+            Contract.ThrowIfNull(semanticModel);
+            if (arrayCreationExpression.Syntax.IsInExpressionTree(semanticModel, expressionType, context.CancellationToken))
+                return;
+
+            var values = arrayCreationExpression.Initializer.ElementValues.SelectAsArray(v => v.ConstantValue.HasValue && v.ConstantValue.Value is not null, v => v.ConstantValue.Value);
+
+            // If we couldn't get constant values for all elements then we can't offer
+            if (values.Length != arrayCreationExpression.Initializer.ElementValues.Length)
+                return;
+
+            var stringValue = GetStringValue(values);
+
+            if (stringValue is null)
+                return;
+
+            var properties = ImmutableDictionary<string, string?>.Empty.Add(StringValuePropertyName, stringValue);
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, arrayCreationExpression.Syntax.GetLocation(), option.Notification.Severity, additionalLocations: null, properties));
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        private static string? GetStringValue(ImmutableArray<object?> values)
+        {
+            try
+            {
+                var byteValues = new byte[values.Length];
+                for (var i = 0; i < values.Length; i++)
+                {
+                    // We shouldn't get nulls, but Convert.ToByte will return (byte)0 if there are any
+                    // so better to crash if the analyzer has a bug, that output a buggy string in users code.
+                    if (values[i] is null)
+                        throw ExceptionUtilities.Unreachable;
+
+                    byteValues[i] = Convert.ToByte(values[i]);
+                }
+
+                var pooledBuilder = PooledStringBuilder.GetInstance();
+                var builder = pooledBuilder.Builder;
+                var ros = new ReadOnlySpan<byte>(byteValues);
+                for (var i = 0; i < ros.Length;)
+                {
+                    // If we can't decode a rune from the array then it can't be represented as a string
+                    if (Rune.DecodeFromUtf8(ros.Slice(i), out var rune, out var bytesConsumed) != OperationStatus.Done)
+                        return null;
+
+                    i += bytesConsumed;
+                    builder.Append(GetStringLiteralRepresentation(rune));
+                }
+
+                return pooledBuilder.ToStringAndFree();
+            }
+            catch
+            {
+                // Ignore any conversion failures and just don't report the diagnostic
+            }
```

Comment:
At which line an exception could happen? I'm concerned about `pooledBuilder` being not freed for such case.

Reply:
i think we can remove the try/catch here.

Reply:
Convert.ToByte will throw, but this was also just being super defensive with arbitrary user data.

The point about the pooled builder is a very good catch though, I completely missed that.

Reply:
note: i would not use Convert.ToByte.  Do we need it?  if we're making a Byte[] full of constants, i would expect them to already be byte-constants.